hudi-spark/src/main/java/org/apache/hudi/keygen/KeyGenerator.java

```diff
@@ -40,4 +40,22 @@ protected KeyGenerator(TypedProperties config) {
    * Generate a Hoodie Key out of provided generic record.
    */
   public abstract HoodieKey getKey(GenericRecord record);
+
+  public abstract String getPartitionPath(GenericRecord record, String partitionPathField);
```

Comment:
I understand you want to reuse the other key generators.. I suggest exposing the methods in those concrete classes without leaking it to the abstract class here.. This will otherwise break all custom key generator implementations out there..

Reply:
Yeah this thing was in my mind at the time of writing this piece of code. So there is a tradeoff here. If you do not want to expose these methods in abstract class, then I cannot initialise keyGenerator in a generic way in getPartitionPath() and getRecordKey() functions in CustomKeyGenerator class like I am doing right now and there will be a bit of redundant code in this class then. Hence I preferred to have it the way it is right now. 

Anyways let me refactor as you are saying, then we can discuss further.